plugins/parade-lspcon/fu-parade-lspcon-device.c

```diff
@@ -0,0 +1,829 @@
+}
+
+static gboolean
+fu_parade_lspcon_probe_active_flash_partition (FuParadeLspconDevice *self,
+					       guint8 *partition,
+					       GError **error)
+{
+	guint8 data = 0x0;
+	g_autoptr(FuParadeLspconI2cAddressGuard) guard = NULL;
+
+	/* read currently-running flash partition number */
+	guard = fu_parade_lspcon_i2c_address_guard_new (self, I2C_ADDR_PAGE5, error);
+	if (guard == NULL)
+		return FALSE;
+	if (!fu_parade_lspcon_read_register (self,
+					     REG_ADDR_ACTIVE_PARTITION,
+					     &data,
+					     error))
+		return FALSE;
+
+	*partition = data;
+	return TRUE;
+}
+
+static gboolean
+fu_parade_lspcon_device_reload (FuDevice *device, GError **error)
+{
+	FuParadeLspconDevice *self = FU_PARADE_LSPCON_DEVICE (device);
+	guint32 oui;
+	guint8 version_buf[2] = { 0x0 };
+	g_autofree gchar *version = NULL;
+	g_autoptr(FuDeviceLocker) aux_device_locker = NULL;
+	g_autoptr(FuUdevDevice) aux_device = NULL;
+	g_autoptr(GList) aux_devices = NULL;
+	g_autoptr(GUdevClient) udev_client = g_udev_client_new (NULL);
+	g_autoptr(GUdevEnumerator) enumerator = g_udev_enumerator_new (udev_client);
+
+	/* determine active partition for flashing later */
+	if (!fu_parade_lspcon_probe_active_flash_partition (self, &self->active_partition, error))
+		return FALSE;
+	g_debug ("device reports running from partition %d", self->active_partition);
+	if (self->active_partition < 1 || self->active_partition > 3) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unexpected active flash partition: %d",
+			     self->active_partition);
+		return FALSE;
+	}
+
+	/* find the drm_dp_aux_dev specified by quirks that is connected to the
+	 * LSPCON, in order to read DPCD from it */
+	if (self->aux_device_name == NULL) {
+		g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no DP aux device specified, unable to query LSPCON");
+		return FALSE;
+	}
+	g_udev_enumerator_add_match_subsystem (enumerator, "drm_dp_aux_dev");
+	g_udev_enumerator_add_match_sysfs_attr (enumerator, "name", self->aux_device_name);
+	aux_devices = g_udev_enumerator_execute (enumerator);
+	if (aux_devices == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to locate a DP aux device named \"%s\"",
+			     self->aux_device_name);
+		return FALSE;
+	}
+	if (g_list_length (aux_devices) > 1) {
+		g_list_free_full (aux_devices, g_object_unref);
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "found multiple DP aux devices with name \"%s\"",
+			     self->aux_device_name);
+		return FALSE;
+	}
+	aux_device = fu_udev_device_new (g_steal_pointer (&aux_devices->data));
+	g_debug ("using aux dev %s", fu_udev_device_get_sysfs_path (aux_device));
+
+	/* the following open() requires the device have IDs set */
+	if (!fu_udev_device_set_physical_id (aux_device, "drm_dp_aux_dev", error))
+		return FALSE;
```

Comment:
Is there any way to split this out? IIRC, i think the new fu-realtek-mst-device.c does something similar?

Reply:
The Realtek version uses the attached I2C tunnel whereas this one reads DPCD directly on the aux device; they're similar in that they need to find a device with given subsystem and sysfs name, but otherwise not.